clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/jobs/JobExecutor.java

```diff
@@ -15,6 +15,33 @@
  */
 package com.netflix.spinnaker.clouddriver.jobs;
 
+import com.netflix.spinnaker.clouddriver.jobs.local.StreamConsumer;
+
+/**
+ * Executes a job defined by a JobRequest, returning the results as a JobStatus.
+ *
+ * The caller can optionally supply a StreamConsumer, in which case the output from the job will be
+ * streamed to the StreamConsumer.
+ *
+ * @see JobRequest
+ * @see JobStatus
+ */
 public interface JobExecutor {
+  /**
+   * Run the specified JobRequest, returning the status and the job's standard output and standard error in
+   * a JobStatus object.
+   * @param jobRequest The job request
+   * @return The result of the job
+   */
   JobStatus runJob(JobRequest jobRequest);
+
+  /**
+   * Runs the specified JobRequest, streaming output to the supplied StreamConsumer, and returns the status and standard
+   * error in a JobStatus object.  The returned JobStatus will have standard output set to null, as the output will have
+   * already been consumed by the supplied consumer.
+   * @param jobRequest The job request
+   * @param streamConsumer A callback that consumes the job's standard output as it is produced
+   * @return The result of the job
+   */
+  JobStatus runJob(JobRequest jobRequest, StreamConsumer streamConsumer);
```

Comment:
If StreamConsumer were changed to something along the lines of a `Function<InputStream, T>` and JobStatus were changed to have a type parameter for its output (which for `JobStatus runJob(JobRequest jobRequest);` would end up being `JobStatus<String>`), the results of the job in this case could be assembled within the function and returned as part of that all the way to the caller of `runJob` as a `JobStatus<List<KubernetesManifest>>` rather than having the `StreamConsumer` lambda close over a mutable collection in the callers scope in order to populate that collection via side effects from whatever call site the lambda gets called at.

That may also be overkill for this, but would likely make certain aspects of testing and future changes a bit more easy by keeping all the various state pretty tightly scoped and with less mutation across method calls

Reply:
I also think an approach like that might make the implementation/code paths of the two methods on this interface more similar, and not necessarily require two separate executeJobStreaming and executeJob methods within `JobExecutorLocal.java`, but I haven't tried that out to be sure

Reply:
I like this idea a lot, and I think it will fix a few of the other comments you've noted below.  I'll give this a try, then will revisit the other comments to see which are still relevant.